src/framework-configuration.js

```diff
@@ -256,17 +303,37 @@ export class FrameworkConfiguration {
   /**
    * Configures an external, 3rd party plugin before Aurelia starts.
    * @param plugin The ID of the 3rd party plugin to configure.
-   * @param config The configuration for the specified plugin.
+   * @param pluginConfig The configuration for the specified plugin.
    * @return Returns the current FrameworkConfiguration instance.
  */
-  plugin(plugin: string, config?: any): FrameworkConfiguration {
+  plugin(
+    plugin: string | ((frameworkConfig: FrameworkConfiguration) => any) | FrameworkPluginInfo,
+    pluginConfig?: any
+  ): FrameworkConfiguration {
     assertProcessed(this);
 
-    if (typeof (plugin) === 'string') {
-      return this.plugin({ moduleId: plugin, resourcesRelativeTo: [plugin, ''], config: config || {} });
+    let info: FrameworkPluginInfo;
+    switch (typeof plugin) {
+    case 'string':
+      info = { moduleId: plugin, resourcesRelativeTo: [plugin, ''], config: pluginConfig || {} };
+      break;
+    case 'function':
+      if (this.queuedPlugins.indexOf(plugin) !== -1) {
+        return this;
+      }
+      this.queuedPlugins.push(plugin);
+      if (this.queuedPlugins.length === 1) {
+        this.postTask(() => this.queuedPlugins = null);
+      } 
+      info = { configure: plugin, config: pluginConfig || {} };
+      break;
+    default:
+      // this is for internal use, from `feature` call
```

Comment:
Mmm... I don't like that because `plugin` is a public API. So if someone passes an invalid value there's no error to help users?

Reply:
I think it's desirable to have plugin and feature resolve to same underlying method / way of registering the plugin / feature information. When refactoring code around this bit, I was (1) keeping things the same way it was, and (2) expecting our type definition should be helpful that devs should not pass anything that is neither a string nor a function here. Is this practical ? We can easily separate them and copy paste the `info =  { ... }` block to feature.